README.md

```diff
@@ -134,10 +134,13 @@ Installing a snap is very quick. Snaps are secure. They are isolated with all of
 
         guix package -i monero
 
-* OS X via [Homebrew](http://brew.sh)
+* macOS via [Homebrew](https://brew.sh/)
 
-        brew tap sammy007/cryptonight
-        brew install monero --build-from-source
+        # install pre-built binaries
+        brew install monero
```

Comment:
Do we want to include this in the official repository? Because Monero is a security software, I think we are highly sensitive about who builds binaries for releases.

Reply:
@sammy007 and @ilovezfs have been, debating, this for a while. I'd like to see consensus or drop all together. Also, I have a PR #3852 demoting the 3rd party packages in favor of source builds or monero binaries.

Reply:
FWIW the Bitcoin repo doesn't seem to have any mention about package-based installations. Maybe we could even drop all third-party packaging info from the official README to make it clear that they're unofficial?

Reply:
>Do we want to include this in the official repository? Because Monero is a security software, I think we are highly sensitive about who builds binaries for releases.

Homebrew is well equipped for this. We have 1.5 million installations of our prebuilt OpenSSL in the last three months alone.

>I'd like to see consensus or drop all together.

Given _anyone_ can set up a third-party tap, it is extremely unusual to have a project like Monero trying to point at someone's third-party tap when the core repository https://github.com/Homebrew/homebrew-core/ already offers the software.


```
        brew tap sammy007/cryptonight
        brew install monero --build-from-source
```
Note I said "trying" because these instructions aren't even correct if the goal is to use sammy007/cryptonight, and will just build the core formula anyway. So you are already pointing at the core formula without realizing it.

Reply:
I can confirm that now my tap contains slightly adjusted formula from brew ofc. I was only against binary bottles because we don't know how they built and also against outdated version after hard protocol upgrade. The link to git repo or tarball in formula is easy to verify and not an issue. Imo it's convenient to stick to brew core and make a **bold clear 3rd party package notice**. I wouldn't personally use any binary bottle from brew, I would check to where it points and build from source.

> We have 1.5 million installations of our prebuilt OpenSSL in the last three months alone.

It's okay until you lose 7 figures due to compromised automated build system which leads to compromised monero private key for instance.

Reply:
I guess the main issue is whether we want to include the binary installation instruction

    brew install monero

in the official README which isn't the case currently. Of course, many users will likely try the binary installation command anyway and find out it works, but I'm not sure if it's a good idea for the project to include the command in the official README which may make it seem like an endorsement.

Reply:
Given your current instructions, any subsequent
```
brew upgrade
```
 or
```
brew upgrade monero
```
or
```
brew reinstall monero
```
has _already_ been using the pre-built bottle anyway, since `--build-from-source` is not a persistent option.

Hence, most of your users are already using the bottles since there have been multiple upgrades since the formula was introduced.

Reply:
@ilovezfs 
> Obviously this is a personal decision, but I think recommending everyone build from source just as a precaution is error prone and opens up additional attack vectors on local user machines.

My personal opinion is exactly the opposite. I don't see at all why building from source would be more error prone and open up more attack vectors than trusting prebuilt binaries.

Reply:
>more error prone

The vast majority of support requests to https://github.com/Homebrew/homebrew-core are caused by someone doing a source build that fails that they then want help debugging.

>open up more attack vectors

compare:
download bottle tarball, brew verifies checksum of tarball, and unpacks tarball
vs.
brew downloads GitHub sources, brew verifies tag, brew compiles locally

In the former we know the bottle matches the checksum checked into source control.

In the latter we only the sources were valid but the build products aren't verified against any checksum.

Reply:
>What if the bottled binaries were compromised?

Then `brew` would refuse to install the bottle because its checksum wouldn't match the SHA-256 in https://github.com/Homebrew/homebrew-core/blob/39c938bdaf4f016b9f7b63aecdbd7b45420a7ae4/Formula/monero.rb#L8-L13

Reply:
I mean, how can one ensure that those binaries against which those SHA-256 checksums have been computed were not compromised? I think this inherently assumes users' trust on Homebrew's build system.

How to distribute secure binaries in a sustainable manner has been a major point of discussion in the Monero community for a long time, and it is expected to be solved by a "reproducible build" system. Until that becomes available, the community will continue to rely on a trusted member (@fluffypony at the moment) building binaries for distribution. I don't want to encourage users to install binaries built by Homebrew.

Reply:
>I think this inherently assumes users' trust on Homebrew's build system.

Yes, using Homebrew's binaries implies that you trust Homebrew's CI build system has not been compromised.

>How to distribute secure binaries in a sustainable manner

Homebrew's initial bottle implementation was added in 2010 and has been distributing prebuilt binaries at scale since 2013. (See https://mikemcquaid.com/2017/09/29/homebrew-ci-evolution/).

>I don't want to encourage users to install binaries built by Homebrew.

As I mentioned above, you're already doing this. The users first `brew upgrade` will get the bottle. Personally I think making that explicit is more honest with your users than saying you should `brew install --build-from-source monero` and then having people ending up with our prebuilt binaries anyway shortly thereafter.

Reply:
In any case, I've updated the PR to remove reference to the prebuilt binaries. Just don't be surprised when nearly everyone installing Monero via Homebrew ends up using the prebuilt due to how people tend to use unqualified `brew upgrade`.

Reply:
> As I mentioned above, you're already doing this.

I wasn't aware of this behavior, thanks for pointing it out. Then the right solution may be to rewrite the current instructions such that the installation of Homebrew's prebuilt binaries is avoided.

Reply:
>I wasn't aware of this behavior, thanks for pointing it out. 

No problem.

>Then the right solution may be to rewrite the current instructions such that the installation of Homebrew's prebuilt binaries is avoided.

For `monero` only or for _everything_ including dependencies?

```
Josephs-MacBook-Pro:~ joe$ brew deps monero
boost
libevent
openssl
readline
unbound
zeromq
Josephs-MacBook-Pro:~ joe$ brew deps monero --include-build
asciidoc
autoconf
automake
boost
cmake
docbook
docbook-xsl
doxygen
gdbm
gettext
gnu-getopt
libevent
libtool
libxml2
makedepend
openssl
pkg-config
python@2
readline
sphinx-doc
sqlite
unbound
xmlto
xz
zeromq
```

Reply:
I guess installing prebuilt binaries except Monero itself is acceptable and practical, though this is subjective. At least I think it's much better than installing Monero prebuilt binaries which directly handle wallet keys.

Reply:
OK. Then that would look like this:

```
To install,
        brew install --build-from-source monero
        brew pin monero

To upgrade,
        brew upgrade
        brew unpin monero
        brew upgrade --build-from-source monero
        brew pin monero
```

Reply:
Personally, my recommendation is to make it just

```
brew install monero
```

and leave all the other details about whether to do a source build or not between the user and the `brew` man page.

Reply:
The command `brew install monero` installs Homebrew's prebuilt binaries, right? If so, that should be best avoided IMO, or at least it shouldn't be endorsed by the official README. I know I can't stop Homebrew from providing prebuilt binaries, but it'd be nice if those binaries get retracted. The complete instructions with `--build-from-source` you posted above seem appropriate, thanks.

Reply:
Homebrew is a binary package manager so retracting the binaries would mean retracting the formula. What may be possible (it would require a policy waiver) would be distributing the upstreaming binaries via https://github.com/Homebrew/homebrew-cask instead of our own binaries https://github.com/Homebrew/homebrew-core. I'm discussing this internally.

That looks like this
```
cask 'monero' do
  version '0.12.2.0'
  sha256 '31568fdedf9557e5d8398013137266afe0e7ed46011691fef1b40ea7cd94e090'

  url "https://dlsrc.getmonero.org/cli/monero-mac-x64-v#{version}.tar.bz2"
  appcast 'https://github.com/monero-project/monero/releases.atom'
  name 'Monero'
  homepage 'https://getmonero.org/'

  binary "#{staged_path}/monero-v#{version}/monero-blockchain-blackball"
  binary "#{staged_path}/monero-v#{version}/monero-blockchain-export"
  binary "#{staged_path}/monero-v#{version}/monero-blockchain-import"
  binary "#{staged_path}/monero-v#{version}/monero-blockchain-usage"
  binary "#{staged_path}/monero-v#{version}/monero-gen-trusted-multisig"
  binary "#{staged_path}/monero-v#{version}/monero-wallet-cli"
  binary "#{staged_path}/monero-v#{version}/monero-wallet-rpc"
  binary "#{staged_path}/monero-v#{version}/monerod"
end
```
The policy in question is that anything open source && command line _must_ build from source and be a formula not a cask, unless it needs signing to function properly (e.g. as is required for access to location services).

Reply:
> Homebrew is a binary package manager

I don't understand why you insist on binaries, homebrew is not only binary package manager, you can use it as gentoo portage.

```
bottle :disable, "You must use official binaries from https://github.com/monero-project/monero/releases".
```

> The policy in question is that anything open source && command line must build from source and be a formula not a cask, unless it needs signing to function properly (e.g. as is required for access to location services).

Once again, `7 figures`. I feel like you don't completely understand what you are packaging. Consider adding to brew policy an exception about permissionless financial network clients, until there is "reproducible build" you don't pour bottles for this type of software.

Reply:
@sammy007 as I said above, if it comes to that, we'll move it to cask. `bottle :disable` is reserved for things that like
```
cryptopp.rb
8:  bottle :disable, "Library and clients must be built on the same microarchitecture"

leafnode.rb
7:  bottle :disable, "leafnode hardcodes the user at compile time with no override available."
```
that break if bottled.

>Once again, 7 figures. I feel like you don't completely understand what you are packaging.

We distribute prebuilt binaries for OpenSSL, GnuPG, and Bitcoin. Monero is not unique.
 You should refrain from telling people they "don't understand" just because you disagree.